services/src/main/java/org/keycloak/authentication/authenticators/directgrant/ValidateRecoveryAuthnCode.java

```diff
@@ -0,0 +1,142 @@
+package org.keycloak.authentication.authenticators.directgrant;
+
+import org.apache.commons.lang.StringUtils;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.AuthenticationFlowError;
+import org.keycloak.authentication.CredentialValidator;
+import org.keycloak.credential.RecoveryAuthnCodesCredentialProvider;
+import org.keycloak.credential.RecoveryAuthnCodesCredentialProviderFactory;
+import org.keycloak.credential.CredentialProvider;
+import org.keycloak.events.Errors;
+import org.keycloak.models.*;
+import org.keycloak.models.credential.RecoveryAuthnCodesCredentialModel;
+import org.keycloak.models.utils.RecoveryAuthnCodesUtils;
+import org.keycloak.provider.ProviderConfigProperty;
+
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+import java.util.LinkedList;
+import java.util.List;
+
+public class ValidateRecoveryAuthnCode extends AbstractDirectGrantAuthenticator
```

Comment:
I don't think we should support recover codes at all with direct-grant, so I would actually suggest removing this one completely?

Reply:
I was wondering about the same yesterday...

However I was thinking about the case when 2-factor is required for browser login and at the same time, you have "Direct Grant" enabled for some clients. Isn't it strange that browser login will require 2-factor, but direct-grant allows to login with simple username/password? Hence it makes sense to me to have authenticator for direct-grant as well (similarly like we have for OTP).

Reply:
This is exactly what we were thinking that it would just be strange not to be able to use recovery codes since you can use it for OTP